x-pack/plugins/lens/public/utils/native_renderer/native_renderer.tsx

```diff
@@ -0,0 +1,57 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import React, { useEffect, useRef } from 'react';
+
+export interface NativeRendererProps<T> {
+  render: (domElement: Element, props: T) => void;
+  actualProps: T;
+  tag?: string;
+  children?: never;
+}
+
+function isShallowDifferent<T>(a: T, b: T): boolean {
```

Comment:
The two loops are nearly identical, could we move the loop into a helper function, and then just return `shallowDiff(a, b) || shallowDiff(b, a)` here? Or, if you're OK with a little allocation, it could be consolidated like so:

```js
function isShallowDifferent(a, b) {
  for (const k of new Set(Object.keys(a || {}).concat(Object.keys(b || {})))) {
    if (!a.hasOwnProperty(k) || !b.hasOwnProperty(k) || a[k] !== b[k]) {
      return true;
    }
  }
  return false;
}
```

Reply:
True, I improved the implementation